### PR TITLE
add e2e tests for replicasets with weight, min and max replicas

### PR DESCRIPTION
**Pull request by jianhuiz**
e2e test with weight, min and max replicas set
#31904 #32014

@quinton-hoole @nikhiljindal @deepak-vij @kshafiee @mwielgus

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32325)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

File: test/e2e/federation-replicaset.go [hunk 1]
```diff
@@ -91,31 +95,148 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
-				waitForReplicaSetOrFail(f.FederationClientset_1_4, f.Namespace.Name, rs.Name, clusters)
+				zeroReplicas := int32(0)
+				rs.Spec.Replicas = &zeroReplicas
+				updateReplicaSetOrFail(f.FederationClientset_1_4, rs)
+				waitForReplicaSetOrFail(f.FederationClientset_1_4, f.Namespace.Name, rs.Name, clusters, nil)
 				f.FederationClientset_1_4.ReplicaSets(f.Namespace.Name).Delete(rs.Name, &api.DeleteOptions{})
+			}
+
+			// general test with default replicaset pref
+			func() {
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: why do you need this to be in func?

Comment (jianhuiz) [inline test/e2e/federation-replicaset.go]: so the cleanup with defer get executed

File: test/e2e/federation-replicaset.go [hunk 2]
```diff
@@ -91,31 +95,148 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	}
+
+	pref = &federation.FederatedReplicaSetPreferences{
+		Clusters: map[string]federation.ClusterReplicaSetPreferences{},
+	}
+	replicas = 0
+	expect = map[string]int32{}
+
+	for i, clusterName := range clusterNames {
+		if i != 0 {
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: why i!=0? you have no guarantee what is on the first position in the array.

File: test/e2e/federation-replicaset.go [hunk 3]
```diff
@@ -91,31 +95,148 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+			clusterRsPref := pref.Clusters[clusterName]
+			clusterRsPref.Weight = int64(1)
+			clusterRsPref.MinReplicas = int64(i + 2)
+			replicas += int32(i + 2)
+			expect[clusterName] = int32(i + 2)
+		}
+	}
+
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: remove empty line

File: test/e2e/federation-replicaset.go [hunk 4]
```diff
@@ -91,31 +95,148 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	replicas += 1
+	expect[clusterNames[0]] = 1
+
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: remove empty line

File: test/e2e/federation-replicaset.go [hunk 5]
```diff
@@ -91,31 +95,148 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+		if i != 0 {
+			clusterRsPref := pref.Clusters[clusterName]
+			clusterRsPref.Weight = int64(100)
+			maxReplicas := int64(i)
+			clusterRsPref.MaxReplicas = &maxReplicas
+			replicas += int32(i)
+			expect[clusterName] = int32(i)
+		}
+	}
+
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: remove empty line

File: test/e2e/federation-replicaset.go [hunk 6]
```diff
@@ -91,31 +95,148 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	replicas += 5
+	expect[clusterNames[0]] = 5
+
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: remove empty line

File: test/e2e/federation-replicaset.go [hunk 7]
```diff
@@ -209,4 +342,12 @@ func newReplicaSet(namespace string, name string, replicas int32) *v1beta1.Repli
 			},
 		},
 	}
+
+	if pref != nil {
+		prefBytes, _ := json.Marshal(pref)
+		prefString := string(prefBytes)
+		rs.Annotations[fedreplicsetcontroller.FedReplicaSetPreferencesAnnotation] = prefString
+	}
+
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: remove empty line

File: test/e2e/federation-replicaset.go (hunk 2 above)
Comment (jianhuiz) [inline test/e2e/federation-replicaset.go]: just want to have a cluster without weight set, don't care which one it is.

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this to confirm if the test is passing

**Comment by jianhuiz** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

File: test/e2e/federation-replicaset.go [hunk 8]
```diff
@@ -94,32 +98,144 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	for i, clusterName := range clusterNames {
+		if i != 0 {
+			clusterRsPref := pref.Clusters[clusterName]
+			clusterRsPref.Weight = int64(1)
+			clusterRsPref.MinReplicas = int64(i + 2)
+			replicas += int32(i + 2)
+			expect[clusterName] = int32(i + 2)
+		}
+	}
+	replicas += 1
```
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: Add a comment saying the extra replica should go to the 0th cluster (since it has the highest weight)

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
We also need to add a test for rebalance

**Comment by nikhiljindal** [comment]
Cluster bring up failed.
Most probably unrelated to this PR.

@k8s-bot federation gce e2e test this

**Comment by quinton-hoole** [comment]
Ping @jianhuiz ?

**Comment by quinton-hoole** [comment]
PS: Please add  "fixes #32014" to your git commit message, so that that issue get automatically closed when this PR merges.

File: test/e2e/federation-replicaset.go [hunk 9]
```diff
@@ -94,32 +98,144 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+			// test for replicaset prefs with weight, min and max replicas
+			createAndUpdateFn := func(pref *federation.FederatedReplicaSetPreferences, replicas int32, expect map[string]int32) {
+				rs := newReplicaSet(nsName, FederationReplicaSetName, replicas, pref)
+				createReplicaSetOrFail(f.FederationClientset_1_4, rs)
+				defer cleanupFn(rs)
+
+				waitForReplicaSetOrFail(f.FederationClientset_1_4, nsName, rs.Name, clusters, expect)
+				By(fmt.Sprintf("Successfuly created and synced replicaset %q/%q (%v/%v) to clusters", nsName, rs.Name, *rs.Spec.Replicas, rs.Status.Replicas))
+
+				rs = newReplicaSet(nsName, FederationReplicaSetName, 0, pref)
```
Comment (jianhuiz) [inline test/e2e/federation-replicaset.go]: it was meant to delete the replicaset (however not supported) for the next test

**Comment by quinton-hoole** [comment]
@mwielgus @nikhiljindal @madhusudancs @matchstick FYI, I have seen some bugs in the replica distribution during manual testing recently.  So I think it would be a good idea to get this PR merged ASAP to help to debug those.

**Comment by nikhiljindal** [comment]
Yes +1.
@jianhuiz Did you get a chance to work on the comments. I dont see a reply from you.
Its expected that the author will reply on the PR when he has fixed all comments, so that reviewers know when it is ready for another round of review.

**Comment by jianhuiz** [comment]
the comments should all have been fixed. I am however unable to trigger the fed e2e test

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
After 20 hours, build logs for federation e2e say "Build Still Running!". There is nothing in artifacts either (just a started.json).
@ixdy @spxtr Any idea what might be happening here?

**Comment by dims** [comment]
cc @kubernetes/test-infra-maintainers

**Comment by jianhuiz** [comment]
rebased

File: test/e2e/federation-replicaset.go (hunk 2 above)
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: Please add the comment.

File: test/e2e/federation-replicaset.go [hunk 10]
```diff
@@ -186,12 +242,131 @@ func verifyCascadingDeletionForReplicaSet(clientset *fedclientset.Clientset, clu
...
+	pref = &federation.FederatedReplicaSetPreferences{
+		Clusters: map[string]federation.ClusterReplicaSetPreferences{
+			clusterNames[0]: {Weight: 1},
+		},
+	}
+	replicas = 0
+	expect = map[string]int32{}
+
+	for i, clusterName := range clusterNames {
+		if i != 0 {
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: same here. It would be good actually if you exclude the i=0 from cluster names.

**Comment by jianhuiz** [comment]
comments fixed

File: test/e2e/federation-replicaset.go [hunk 11]
```diff
@@ -186,12 +242,131 @@ func verifyCascadingDeletionForReplicaSet(clientset *fedclientset.Clientset, clu
 	}
 }
 
-func waitForReplicaSetOrFail(c *fedclientset.Clientset, namespace string, replicaSetName string, clusters map[string]*cluster) {
-	err := waitForReplicaSet(c, namespace, replicaSetName, clusters)
+func generageFedRsPrefsWithWeight(clusters map[string]*cluster) (pref *federation.FederatedReplicaSetPreferences, replicas int32, expect map[string]int32) {
+	clusterNames := make([]string, 0, len(clusters))
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: Please generate this list in one place (init, helper function or so)

File: test/e2e/federation-replicaset.go [hunk 12]
```diff
@@ -186,12 +242,131 @@ func verifyCascadingDeletionForReplicaSet(clientset *fedclientset.Clientset, clu
 	}
 }
 
-func waitForReplicaSetOrFail(c *fedclientset.Clientset, namespace string, replicaSetName string, clusters map[string]*cluster) {
-	err := waitForReplicaSet(c, namespace, replicaSetName, clusters)
+func generageFedRsPrefsWithWeight(clusters map[string]*cluster) (pref *federation.FederatedReplicaSetPreferences, replicas int32, expect map[string]int32) {
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: generate

File: test/e2e/federation-replicaset.go [hunk 13]
```diff
@@ -186,12 +242,131 @@ func verifyCascadingDeletionForReplicaSet(clientset *fedclientset.Clientset, clu
...
+			clusterRsPref.Weight = int64(i)
+			replicas += int32(i)
+			expect[clusterName] = int32(i)
+		}
+	}
+
+	return
+}
+
+func generageFedRsPrefsWithMin(clusters map[string]*cluster) (pref *federation.FederatedReplicaSetPreferences, replicas int32, expect map[string]int32) {
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: generate

File: test/e2e/federation-replicaset.go [hunk 14]
```diff
@@ -186,12 +242,131 @@ func verifyCascadingDeletionForReplicaSet(clientset *fedclientset.Clientset, clu
...
+	clusterNames := make([]string, 0, len(clusters))
```
Comment (mwielgus) [inline test/e2e/federation-replicaset.go]: same as above

**Comment by mwielgus** [comment]
minor nits, please rebase.

**Comment by jianhuiz** [comment]
comment filxed and rebased

**Comment by mwielgus** [comment]
LGTM, please rebase and I will apply the label.

**Comment by nikhiljindal** [comment]
@k8s-bot kops aws e2e test this
@k8s-bot node e2e test this

**Comment by nikhiljindal** [comment]
@jianhuiz Please rebase.
Looks like the last time you rebased, you didnt comment here.
Without a comment, Marcin or I dont get a notification that you have rebased.
Please rebase and comment here.

Apologies that this has been pending for so long

**Comment by jianhuiz** [comment]
rebased

**Pull request merged**